src/main/java/mate.academy/lib/Injector.java

```diff
@@ -1,13 +1,85 @@
 package mate.academy.lib;
 
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.util.HashMap;
+import java.util.Map;
+import mate.academy.service.FileReaderService;
+import mate.academy.service.ProductParser;
+import mate.academy.service.ProductService;
+import mate.academy.service.impl.FileReaderServiceImpl;
+import mate.academy.service.impl.ProductParserImpl;
+import mate.academy.service.impl.ProductServiceImpl;
+
 public class Injector {
-    private static final Injector injector = new Injector();
+    private static final Map<Class<?>, Class<?>> interfacesImplementation;
+    private static final Injector injector;
+    private Map<Class<?>, Object> instances;
+
+    public Injector() {
+        instances = new HashMap<>();
+    }
+
+    static {
+        interfacesImplementation = Map.of(
+                FileReaderService.class, FileReaderServiceImpl.class,
+                ProductParser.class, ProductParserImpl.class,
+                ProductService.class, ProductServiceImpl.class
+        );
+        injector = new Injector();
+    }
 
     public static Injector getInjector() {
         return injector;
     }
 
     public Object getInstance(Class<?> interfaceClazz) {
-        return null;
+        if (!interfaceClazz.isAnnotationPresent(Component.class)) {
+            throw new RuntimeException("Class must be annotated with @Component");
+        }
+        Object newInstance = null;
+        Class<?> clazz = findImplementation(interfaceClazz);
+        Field[] declaredFields = clazz.getDeclaredFields();
+        for (Field field : declaredFields) {
+            if (field.isAnnotationPresent(Inject.class)) {
+                Object fieldInstance = getInstance(field.getType());
+                newInstance = createNewInstance(clazz);
+                try {
+                    field.setAccessible(true);
+                    field.set(newInstance, fieldInstance);
+                } catch (IllegalAccessException e) {
+                    throw new RuntimeException("Can't initialize field value. "
+                            + "Class: " + clazz.getName() + ". "
+                            + "Field: " + field.getName());
+                }
+            }
+        }
+        if (newInstance == null) {
+            newInstance = createNewInstance(clazz);
+        }
+        return newInstance;
+    }
+
+    private Object createNewInstance(Class<?> clazz) {
+        if (instances.containsKey(clazz)) {
+            return instances.get(clazz);
+        }
+        try {
+            Constructor<?> constructor = clazz.getConstructor();
+            Object instance = constructor.newInstance();
+            instances.put(clazz, instance);
+            return instance;
+        } catch (InstantiationException
+                 | InvocationTargetException
+                 | NoSuchMethodException
+                 | IllegalAccessException e) {
```

Comment:
It is better to replace many exceptions that have a common parent with a general parental exception.

Bad example: 
} catch (InvocationTargetException | InstantiationException
      | IllegalAccessException | NoSuchMethodException e) {

Good example: 
} catch (ReflectiveOperationException e) {